### PR TITLE
Ios actionable notifications

### PR DESCRIPTION
**Pull request by aurimasc**
Adds support for iOS actionable notifications.
Add API to register notification categories with associated actions and to get which action was used and input.
Actionable notifications can be dragged down to reveal the actions.
The test project update to have a couple buttons with simple buttons and text input.

### CODE REVIEW

**Comment by vaidasma** [comment]
Both action types are present in the test project and are able to work
• Correct choice is returned back to the app
• Correct input is returned back to the app

Issues found:
• Clearing actionable notification and then trying to interact with another actionable notification sent right after does not work on iPhone 6s (both actionable types from the test project). Clicking any action does nothing
• If the app is in foreground, the actionable notifications do not arrive to the device, but the app registers that it arrived
• In test project, the actionable choices should be:  Respond / Reject (not Response / Reject), to have consistency of actions, not a verb and noun

**Review by vaidasma — APPROVED** [review]
Found issues have been fixed

**Review by todi1856 — COMMENTED** [review]
So, I don't have too much experience with notifications, but GetLast* API seems questionable.

What happens if user performs more than one notification action per frame, it seems like there's no way to react to all of them?

File: com.unity.mobile.notifications/Runtime/iOS/Plugins/UnityNotificationWrapper.m [hunk 1]
```diff
@@ -173,4 +173,127 @@ bool _GetAppOpenedUsingNotification()
...
+    NSString* titl = [NSString stringWithUTF8String: title];
+    NSString* btnTitle = [NSString stringWithUTF8String: buttonTitle];
+    NSString* placeHolder = placeholder ? [NSString stringWithUTF8String: placeholder] : NULL;
+    UNTextInputNotificationAction* action = [UNTextInputNotificationAction actionWithIdentifier: idr title: titl options: opts textInputButtonTitle: btnTitle textInputPlaceholder: placeHolder];
+    return (__bridge_retained void*)action;
+}
+
+void _ReleaseUNNotificationAction(void* action)
+{
+    UNNotificationAction* a = (__bridge_transfer UNNotificationAction*)action;
```
[the remark this answers is not in the document]
Comment (aurimasc) [reply com.unity.mobile.notifications/Runtime/iOS/Plugins/UnityNotificationWrapper.m:216]: You mean the assignment below? It's not actually needed, just that I don't like unused variable and very much don't like casting and not doing anything with result, so sch assignment seams to be least bad.
bridge retain/transfer casts are for ref count management when interoping with C.

**Comment by aurimasc** [comment]
> So, I don't have too much experience with notifications, but GetLast* API seems questionable.
> 
> What happens if user performs more than one notification action per frame, it seems like there's no way to react to all of them?

It's not humanly possible to perform more than one, because you click a button on a visual notification, which brings you back to application (which is paused while you perform the action).

**Comment by todi1856** [comment]
> > So, I don't have too much experience with notifications, but GetLast* API seems questionable.
> > What happens if user performs more than one notification action per frame, it seems like there's no way to react to all of them?
> 
> It's not humanly possible to perform more than one, because you click a button on a visual notification, which brings you back to application (which is paused while you perform the action).

Hmph, not sure, it still feels that there's at least some possibility for this to happen, like when the application is super laggy, etc. So API like that is questionable. Can I ask, why you didn't choose to use callbacks instead which a more safer in this case?

**Comment by aurimasc** [comment]
> Can I ask, why you didn't choose to use callbacks instead which a more safer in this case?

Notifications can be responded to when app is not even running. Responding to notification then will launch the app. In ObjC we have register on launch to receive response, but users code is not running at this point. So either way we have to cache received actions.
Either way, if you care about responded notifications or actions, you have to do it in Start()/OnApplicationPause(<true>), since interaction with notifications causes your app to pause.

**Review by todi1856 — APPROVED** [review]
Approving for now. Maybe show it to @vincent-unity as well, just to have more eyes on this?

**Pull request merged**